app/actions/ControlActions.js

```diff
@@ -924,3 +943,171 @@ export const getPeerInfo = () => (dispatch, getState) => {
+
+const setAccountPassphrase = async (
+  walletService,
+  accountNumber,
+  accountPassphrase,
+  newAcctPassphrase,
+  walletPassphrase
+) => {
+  try {
+    await wallet.setAccountPassphrase(
+      walletService,
+      accountNumber,
+      accountPassphrase,
+      newAcctPassphrase,
+      walletPassphrase
+    );
+    return;
+  } catch (error) {
+    return error;
+  }
+};
+
+export const setAccountsPass = (walletPassphrase) => async (
+  dispatch,
+  getState
+) => {
+  dispatch({ type: SETACCOUNTSPASSPHRASE_ATTEMPT });
+  try {
+    const oldAccounts = sel.balances(getState());
+    await wallet.unlockWallet(sel.walletService(getState()), walletPassphrase);
+    const accounts = Promise.all(
+      oldAccounts.map(async (acct) => {
+        // just skip if imported account.
+        if (acct.accountNumber === Math.pow(2, 31) - 1) {
+          return acct;
+        }
+        // we set the account passphrase as the wallet passphrase to avoid the user
+        // ending with multiple passphrases.
+
+        await setAccountPassphrase(
+          sel.walletService(getState()),
+          acct.accountNumber,
+          walletPassphrase,
+          walletPassphrase,
+          walletPassphrase
+        );
+        acct.encrypted = true;
+        return acct;
+      })
+    );
+    return dispatch({ type: SETACCOUNTSPASSPHRASE_SUCCESS, accounts });
+  } catch (error) {
+    dispatch({ error, type: SETACCOUNTSPASSPHRASE_FAILED });
+  }
+};
+
+// unlock
+export const UNLOCKACCOUNT_ATTEMPT = "UNLOCKACCOUNT_ATTEMPT";
+export const UNLOCKACCOUNT_FAILED = "UNLOCKACCOUNT_FAILED";
+export const UNLOCKACCOUNT_SUCCESS = "UNLOCKACCOUNT_SUCCESS";
+// lock
+export const LOCKACCOUNT_ATTEMPT = "LOCKACCOUNT_ATTEMPT";
+export const LOCKACCOUNT_FAILED = "LOCKACCOUNT_FAILED";
+export const LOCKACCOUNT_SUCCESS = "LOCKACCOUNT_SUCCESS";
+
+// unlockAcctAndExecFn unlocks the account and performs some action. Locks the
+// account in case of success or error, if leaveUnlock is not informed.
+export const unlockAcctAndExecFn = (
+  passphrase,
+  acctNumber,
+  fn,
+  leaveUnlock
+) => async (dispatch, getState) => {
+  let res = null;
+  let fnError = null;
+  dispatch({ type: UNLOCKACCOUNT_ATTEMPT });
+
+  // sanity checks
+  const accounts = sel.balances(getState());
+
+  const account = accounts.find((acct) => acct.accountNumber === acctNumber);
+  if (!account) {
+    throw "Account not found";
+  }
+  if (!account.encrypted) {
+    throw "Account not encrypted";
+  }
+
+  // unlock wallet
+  try {
+    await wallet.unlockAccount(
+      sel.walletService(getState()),
+      passphrase,
+      acctNumber
+    );
+  } catch (error) {
+    // no need to lock as unlock errored.
+    dispatch({ type: UNLOCKACCOUNT_FAILED, error });
+    throw error;
+  }
+
+  // execute method
+  try {
+    res = await fn();
+  } catch (error) {
+    fnError = error;
+  }
+
+  if (leaveUnlock) {
+    return res;
```

Comment:
You also need to check for `fnError` here, otherwise you could return `res` when the actual `fn()` call failed.